File: 01-conversation-loop/src/anthropic.ts
/**
 * 真实 API 调用
 * 支持 Anthropic 官方 API 或兼容接口（如智谱）
 *
 * 配置方式（二选一）：
 *   1. 项目根目录 .env 文件
 *   2. 环境变量 ANTHROPIC_API_KEY + ANTHROPIC_BASE_URL
 */

import Anthropic from "@anthropic-ai/sdk"
import { readFileSync } from "node:fs"
import { resolve } from "node:path"
import type { Message, ContentBlock, Tool } from "./types.js"

// 加载 .env 文件（不依赖 dotenv 库）
function loadEnv() {
  const envPath = resolve(process.cwd(), ".env")
  try {
    const content = readFileSync(envPath, "utf-8")
    for (const line of content.split("\n")) {
      const trimmed = line.trim()
      if (!trimmed || trimmed.startsWith("#")) continue
      const eqIndex = trimmed.indexOf("=")
      if (eqIndex === -1) continue
      const key = trimmed.slice(0, eqIndex).trim()
      const value = trimmed.slice(eqIndex + 1).trim()
      if (!process.env[key]) {
        process.env[key] = value
      }
    }
  } catch {
    // .env 文件不存在，使用环境变量
  }
}

loadEnv()

export function createAnthropicCaller(model: string = "claude-sonnet-4-20250514") {
  const apiKey = process.env.ANTHROPIC_API_KEY
  if (!apiKey) {
    throw new Error("请设置 ANTHROPIC_API_KEY（环境变量或 .env 文件）")
  }

  const baseURL = process.env.ANTHROPIC_BASE_URL

  const client = new Anthropic({
    apiKey,
    ...(baseURL ? { baseURL } : {}),
  })

  console.log(`[API] 基地址: ${baseURL || "https://api.anthropic.com"}`)
  console.log(`[API] 模型: ${model}`)

  return async function callAnthropic(
    messages: Message[],
    tools: Tool[],
  ): Promise<ContentBlock[]> {
    const sdkMessages = messages.map((msg) => ({
      role: msg.role,
      content: msg.content.map((block) => {
        if (block.type === "text") return { type: "text" as const, text: block.text }
        if (block.type === "tool_use") {
          return {
            type: "tool_use" as const,
            id: block.id,
            name: block.name,
            input: block.input,
          }
        }
        if (block.type === "tool_result") {
          return {
            type: "tool_result" as const,
            tool_use_id: block.tool_use_id,
            content: block.content,
            is_error: block.is_error,
          }
        }
        return block
      }),
    }))

    const response = await client.messages.create({
      model,
      max_tokens: 4096,
      messages: sdkMessages as any[],
      tools: tools.map((t) => ({
        name: t.name,
        description: t.description,
        input_schema: t.input_schema as any,
      })),
    })

    return response.content.map((block): ContentBlock => {
      if (block.type === "text") {
        return { type: "text", text: block.text }
      }
      if (block.type === "tool_use") {
        return {
          type: "tool_use",
          id: block.id,
          name: block.name,
          input: block.input as Record<string, unknown>,
        }
      }
      return { type: "text", text: JSON.stringify(block) }
    })
  }
}

File: 06-rust-rewrite/src/anthropic.rs
//! Anthropic API 调用器
//!
//! 对照 TS 版 anthropic.ts
//! 调用真实的 Claude API，支持工具调用

use async_trait::async_trait;
use serde_json::{json, Value};

use crate::model::ModelCaller;
use crate::tools::ToolRegistry;
use crate::types::{ChatMessage, ContentBlock};

/// Anthropic API 调用器
///
/// 对照 TS 版的 createAnthropicCaller()
/// 差异：Rust 用 struct + trait 替代闭包工厂
pub struct AnthropicCaller {
    client: reqwest::Client,
    api_key: String,
    base_url: String,
    model: String,
    tool_schemas: Vec<Value>,
    system_prompt: String,
}

impl AnthropicCaller {
    /// 从环境变量读取配置，从注册表提取工具定义
    ///
    /// 对照 TS 版：new Anthropic({ apiKey, baseURL })
    pub fn new(registry: &ToolRegistry) -> Result<Self, String> {
        let api_key = std::env::var("ANTHROPIC_API_KEY")
            .map_err(|_| "请设置 ANTHROPIC_API_KEY（.env 文件或环境变量）".to_string())?;
        let base_url = std::env::var("ANTHROPIC_BASE_URL")
            .unwrap_or_else(|_| "https://open.bigmodel.cn/api/anthropic".into());
        let model = std::env::var("ANTHROPIC_MODEL")
            .unwrap_or_else(|_| "glm-5.1".into());

        // 提取工具 schema——对照 TS 版 tools.map(t => ({ name, description, input_schema }))
        let tool_schemas: Vec<Value> = registry
            .all_tools()
            .iter()
            .map(|t| {
                json!({
                    "name": t.name(),
                    "description": t.description(),
                    "input_schema": t.input_schema()
                })
            })
            .collect();

        let tools = registry.all_tools();
        let tool_names: Vec<String> = tools.iter().map(|t| t.name().to_string()).collect();
        let system_prompt = format!(
            "你是一个代码分析助手，可以使用以下工具来帮助用户：{}。\n请用中文回复。",
            tool_names.join("、")
        );

        Ok(Self {
            client: reqwest::Client::new(),
            api_key,
            base_url,
            model,
            tool_schemas,
            system_prompt,
        })
    }

    pub fn model_name(&self) -> &str {
        &self.model
    }

    /// 将内部消息格式转换为 Anthropic API 格式
    ///
    /// 对照 TS 版：messages.map(m => ({ role, content }))
    fn convert_messages(&self, messages: &[ChatMessage]) -> Vec<Value> {
        messages
            .iter()
            .map(|msg| {
                let content: Vec<Value> = msg
                    .content
                    .iter()
                    .map(|block| match block {
                        ContentBlock::Text { text } => {
                            json!({"type": "text", "text": text})
                        }
                        ContentBlock::ToolUse { id, name, input } => {
                            json!({
                                "type": "tool_use",
                                "id": id,
                                "name": name,
                                "input": input
                            })
                        }
                        ContentBlock::ToolResult {
                            tool_use_id,
                            content,
                            ..
                        } => {
                            json!({
                                "type": "tool_result",
                                "tool_use_id": tool_use_id,
                                "content": content
                            })
                        }
                    })
                    .collect();
                json!({"role": msg.role, "content": content})
            })
            .collect()
    }
}

#[async_trait]
impl ModelCaller for AnthropicCaller {
    async fn call(&self, messages: &[ChatMessage]) -> Vec<ContentBlock> {
        let api_messages = self.convert_messages(messages);

        let body = json!({
            "model": self.model,
            "max_tokens": 4096,
            "system": self.system_prompt,
            "messages": api_messages,
            "tools": self.tool_schemas,
        });

        let url = format!("{}/v1/messages", self.base_url);

        // 发送请求——对照 TS 版 client.messages.create()
        let resp = match self
            .client
            .post(&url)
            .header("x-api-key", &self.api_key)
            .header("anthropic-version", "2023-06-01")
            .json(&body)
            .send()
            .await
        {
            Ok(r) => r,
            Err(e) => {
                return vec![ContentBlock::Text {
                    text: format!("[API连接失败: {}]", e),
                }]
            }
        };

        let status = resp.status();
        let text = resp.text().await.unwrap_or_default();

        let data: Value = match serde_json::from_str(&text) {
            Ok(d) => d,
            Err(_) => {
                return vec![ContentBlock::Text {
                    text: format!("[API响应解析失败: {}]", &text[..text.len().min(100)]),
                }]
            }
        };

        // 解析 content 数组
        let Some(content) = data["content"].as_array() else {
            return vec![ContentBlock::Text {
                text: format!(
                    "[API错误 {}: {}]",
                    status,
                    data["error"]["message"].as_str().unwrap_or("未知错误")
                ),
            }];
        };

        content
            .iter()
            .filter_map(|b| match b["type"].as_str() {
                Some("text") => Some(ContentBlock::Text {
                    text: b["text"].as_str().unwrap_or("").to_string(),
                }),
                Some("tool_use") => Some(ContentBlock::ToolUse {
                    id: b["id"].as_str().unwrap_or("").to_string(),
                    name: b["name"].as_str().unwrap_or("").to_string(),
                    input: b["input"].clone(),
                }),
                _ => None,
            })
            .collect::<Vec<_>>()
    }
}
